experimental/inductor.py

```diff
@@ -0,0 +1,669 @@
+        default_shard_schema = Schema(mesh, [Replicate()])
+    else:
+        default_shard_schema = None
+
+    with FakeTensorMode():
+        mc = MyModule()
+        sgd = optim_cls(mc.parameters(), lr=0.01)
+
+        fake_x = torch.ones(10, 4)
+        fake_y = torch.ones(10, 2)
+
+        if DIST:
+            mark_placements(fake_x, [Shard(0)])
+            mark_placements(fake_y, [Shard(0)])
+
+            # mark_placements(mc.linear.weight, [Shard(0]])
+
+        # warning: even though we'ree passing mc and sgd as "example" inputs,
+        #    the tracer actually takes its tensor states and saves them as part of
+        #    the compiled_f. So mc and sgd cannot be mutated directly afterwards.
+        compiled_f = make_inductor_fn(train_step, [mc, sgd, fake_x, fake_y], default_shard_schema=default_shard_schema)
+
+    if DIST:
+        # make sure our parameter is the same across devices initially
+        # doing this after compile because during compilation we're running the training step a few times
+        # TODO: fix this. we should run compilation with fake tensors.
+        # for p in mc.parameters(): 
+        #    torch.distributed.all_reduce(p)
+        x = distribute_tensor(x, mesh, placements=[Shard(0)])
+        y = distribute_tensor(y, mesh, placements=[Shard(0)])
+
+    with torch.no_grad():
+        for i in range(10): 
+            # FIXME attention -- mc and sgd are actualy _not_ being used!!
+            #    this is because we collected parameter/tensors from mc and sgd
+            #    during the first tracing, and we're reusing those here instead of
+            #    actually taking them from mc and sgd.
+            loss_shard,  = compiled_f(mc, sgd, x, y)
+            dprint(f'iteration: local_loss={loss_shard}')
+
+
+
+class TrackingMaterializer:
+    def __init__(self, materializer):
+        self.map: Dict[torch.Tensor, torch.Tensor] = dict()
+        self.materializer = materializer
+
+    def __call__(self, tensor, param=None):
+        if tensor in self.map:
+            return self.map[tensor]
+        elif param is not None and param in self.map:
+            return self.map[param]
+        else:
+            materialized = self.materializer(tensor)
+            self.map[tensor] = materialized
+            self.map[param] = materialized
+            return materialized
+
+
+
+def main_fake(optim_cls):
+    fake_mode = FakeTensorMode()
+    torch.set_default_device('cuda')
+
+    # with fake_mode:
+    with fake_mode:
+        # 1.1. create fake model
+        mc = MyModule()
+
+        # 1.2. create fake optimizers
+        sgd = optim_cls(mc.parameters(), lr=0.01, capturable=True)
+
+        # 1.3. create fake input examples
+        fake_x = torch.ones(10, 4)
+        fake_y = torch.ones(10, 2)
+
+        compiled_f = make_inductor_fn(train_step, [mc, sgd, fake_x, fake_y], default_shard_schema=None)
+
+    # 2.3. create real input examples (should be DTensors)
+    x = torch.ones(10, 4)
```

Comment:
Should we have called distribute_tensor() on x and y before running compiled_f?

Reply:
Yes this is actually done, maybe this is an older version of the diff?